### ISSUE TITLE
Feature request: Detect/notify on port changes

### ISSUE THREAD

**Issue by mcclure** [thestk/rtmidi#191]
I am attempting to deal with what happens when midi devices hot swap at runtime. I am building my own mapping of device id -> current port number so that if a device goes away and comes back I know it is the same device. The problem is I do not know when I need to rebuild my map. It is not clear at what times RTMidi's port view can change and I do not know how to detect a change except by iterating all ports and comparing the strings, which seems slightly wasteful.

I would like to request either some kind of a call like >havePortsChanged(), or a message that gets queued with midiin->getMessage which flags ports have changed. RTMidi is in a better position to determine when a port change has occurred than my client code is.

If this is not feasible, something that would help is better documentation around port changes. It would be very helpful to be able to know at what times ports can change or when I need to a port change. If I need to potentially be ready for ports to change between any two calls\*, it would be useful to have that fact documented. Everything around ports is very unclear in the current docs, for example, although its usage is clear from the docs I cannot find an explanation anywhere of what openPort(0) actually *does*.

I am aware of #30 but I am slightly confused what to do with it. It looks like an entirely different fork with things like different build scripts and namespacing. This is a smaller, more attainable question. Since #30 has not been integrated over a period of five years maybe it is worth exploring smaller solutions.

-------

\* I did a test on OS X 10.13.6 which seems to indicate that port assignments can change at any time as if the change were occurring in a simultaneous thread. Perhaps this is the race condition other issues refer to. I ran:

    int outc = midiOut.getPortCount();
    int inc = midiIn.getPortCount();
    ostringstream str;
    str << "Out " << outc << endl;
    for(int c = 0; c < outc; c++) {
    	str << c << ": " << midiOut.getPortName(c) << endl;
    }
    str << "In " << inc << endl;
    for(int c = 0; c < inc; c++) {
    	str << c << ": " << midiIn.getPortName(c) << endl;
    }

If I run this I frequently see

    MidiInCore::getPortName: the 'portNumber' argument (1) is invalid.

This implies port 1 disappeared between midiIn.getPortCount() and midiIn.getPortName().

**Comment by garyscavone** [comment]
This is an ongoing and unfixed issue primarily because there is no simple solution (that I am aware of) that can work across all supported APIs. I think there is a callback with CoreMidi to tell us when a device connection changes but as far as I remember, there was no such thing for Windows (but I admit to not looking into the Windows API in a while). If someone can find a way to implement something like a “deviceChangeCallback”, I’m happy to support the effort.

Given this problem, it is best to organize your code in such a way that it does not depend on devices always staying connected. For example, a device selection popup should only appear when the user wants to establish a connection. Obviously, a user should know that if they disconnect a device while it is in use, then the program behaviour might be a bit problematic.

As for your example at the bottom, I’m surprised you are getting that behaviour. My own use with CoreMidi has been pretty stable. Did you open/close programs that establish virtual ports (or connect / disconnect physical MIDI devices) while running that? The same thing happens in the midiProbe.cpp example. Does that produce the same result?

**Comment by mcclure** [comment]
I'm sorry, I messed up typing the above, I saw the `MidiInCore::getPortName: the 'portNumber' argument (1) is invalid` error while running the included code in a loop and repeatedly plugging in and unplugging a MIDI device. Is this still surprising? I can test with midiprobe but that of course does not run in a loop.

"it is best to organize your code in such a way that it does not depend on devices always staying connected"

The problem with this approach is that say I have two devices, and I am sending MIDI data to the device on port 0, and then the device at port 0 is disconnected. Now the device that was on port 1 is on port 0 and I am sending data to the wrong device. So it seems to me that literally all RtMIDI code "depends on devices always staying connected" or, at least, depends on being able to detect that a device has disconnected.

**Comment by garyscavone** [comment]
Yes, if you ran the code in a loop and repeatedly plug and unplug devices, I can definitely see how you would get that error. 

The device enumeration used by RtMidi is essentially what all the MIDI APIs use themselves. They simply enumerate a list of devices, the order of which is somewhat undefined. Then you use the device number to open a port. So, the APIs do not provide a mechanism (generally) that gets around the problem of having a device plugged or unplugged while being enumerated or used.  But don’t you think that the user should be aware that this will potentially cause problems? If you are sending MIDI data to the device on port 0 and then someone unplugs it, that should be expected to produce undefined behaviour. I think the best scenario would be that you get an error from the system saying that the connection is lost and then you respond to that by telling the user they did something they shouldn’t have done. And I believe RtMidi will throw an error in that case.

**Comment by mcclure** [comment]
If that is the case it would still be helpful to explicitly explain this in the documentation, as currently it is not clear that a device could disappear between calling getPortCount() and acting on it on the next line.

**Comment by garyscavone** [comment]
I just added a few sentences to the documentation about this:

======
Note that the port enumeration is system specific and will change if any devices are unplugged or plugged (or a new virtual port opened or closed) by the user. Thus, the port numbers should be verified immediately before opening a stream. As well, if a user unplugs a device (or closes a virtual port) while a port connection exists to that device/port, a MIDI system error will be generated.
======

It should be expanded but hopefully that at least gets things rolling in this direction.

**Comment by mcclure** [comment]
That helps!

**Comment by iankiller77** [comment]
It might help a little, but for windows and mac you can identify a connection/disconnection by testing the handles. In the case of a disconnected device, trying to get the message would give a invalid handler error on windows. Don't know the code for MAC but it's work pretty much the same.

You can do the same for sending message, you will get a invalid handle message. I didn't look into linux much considering I didn't have to use it yet.

A little side note, some Synthesizer doesn't invalidate handle properly on shut down if the port was active. It's the case for VirtualMidiSynth on windows. I think that the issue happen only for Synth close to the OS. (VirtualMidiSynth use it's own driver)

A quick solution to validate a port would simply be to send a DUMMY message to the port from time to time. Having an error doing so will most likely mean that your port is invalidated.

**Comment by SpotlightKid** [comment]
FWIW, Jack also allows to register callbacks for when new clients or ports are registered or ports are renamed:

http://jackaudio.org/files/docs/html/group__ClientCallbacks.html

**Comment by mcclure** [comment]
If CoreMIDI and Jack support this feature but it turns out Windows does not, it seems like it would be possible for RtMidi to emulate the feature in a degraded way on Windows (ie a thread wakes up every 1ms and polls devices) and there would be significant benefit even if the Windows version didn't work as well. As with #195 this is something that RtMidi is more able to implement than user code because it is talking to the OS MIDI services directly.

**Comment by atsushieno** [comment]
ALSA does not support it either. I tried to implement this feature in my own C# library but gave up for the same reason. If you read Chromium source code for Web MIDI support, you will find that it is quite complicated.

If you are really serious about connection state change detection, run a loop like once in a few seconds to get list of devices and report the changes. That is ugly and I wouldn't implement such code for my library and wouldn't expect rtmidi to support that either...

**Comment by v1ne** [comment]
About Windows: One way to get notified whenever some MIDI port is added or removed is by taking a close look at `WM_DEVICECHANGE` messages.

And just for comparison: JUCE uses polling for changes to MIDI interfaces in its `AudioDeviceSelectorComponent`, on all platforms. ;)

**Comment by thebne** [comment]
Just noticed that. Is there a way to access the underlying port/device in Linux? I'd want to use notify/select in order to mitigate it

**Comment by madbrain76** [comment]
I ran into this issue with Grandorgue on Linux, which uses the Rtmidi Alsa backend. Someone else ran into it too on Windows.
https://github.com/GrandOrgue/grandorgue/discussions/1186

If I understand correctly, one reason this hasn't been implemented is that this feature is not consistently available across all operating systems and MIDI APIs. If that's the main holdup, might it make sense to implement it piece-meal, for the backends that do have this ability ? I'm letting Codex take a stab at doing this for my own Alsa / Linux use case.

**Comment by insolace** [comment]
This would be a major api extension so it’s likely not in scope for rtmidi. However, we implemented this using rtmidi in our product editors, and we recently open sourced it. Take a look at kmi_ports.cpp/h in, codex can probably port this into whatever you need.GitHub - Muse-Kinetics/kmi_midi_device_manager: KMI MIDI Device Manager - Cross-platform C++/Qt library for KMI devicesgithub.comOn Sep 5, 2026, at 2:48 AM, Julien Pierre ***@***.***> wrote:﻿madbrain76 left a comment (thestk/rtmidi#191)
I ran into this issue with Grandorgue on Linux, which uses the Rtmidi Alsa backend. Someone else ran into it too on Windows.
GrandOrgue/grandorgue#1186
If I understand correctly, one reason this hasn't been implemented is that this feature is not consistently available across all operating systems and MIDI APIs. If that's the main holdup, might it make sense to implement it piece-meal, for the backends that do have this ability ? I'm letting Codex take a stab at doing this for my own Alsa / Linux use case.

**Comment by madbrain76** [comment]
> This would be a major api extension so it’s likely not in scope for rtmidi. 

I'll see what the maintainers say about my PR #374  . It does not change the public API for clients, only the internal implementation.

>However, we implemented this using rtmidi in our product editors, and we recently open sourced it. Take a look at kmi_ports.cpp/h in, codex can probably port this into whatever you need.

Your link was broken, but I found your library on your profile. It is a different API, though. The only program I need the automation reconnection feature for is Grandorgue, which is built on rtmidi. It would likely be a big change to add support for kmi into Grandorgue. And I'm just a user, not a maintainer. If you look at the GO discussion I linked to above, you'll see more - there is a Microsoft employee discussing reconnection. Coreaudio on MacOS reportedly has solved this at the OS level, also. My personal use case is with Linux, though I can test on Windows as well.
